truex-ios-mobile-reference-app/UnlockSoloViewController.m

```diff
@@ -0,0 +1,313 @@
+//- (void)safariViewControllerDidFinish:(SFSafariViewController *)controller {
+//    // true[X] - You will need remume the true[X] Ad Renderer after safariViewController
+//    if (self.activeAdRenderer) {
+//        [self.activeAdRenderer resume];
+//    }
+//}
+
+
+// MARK: - Helper Functions / Fake Ad Framework
+- (void)fetchAd:(NSString *)url {
+    if (self.vastDictionary != nil) {
+        return;
+    }
+    url = [url stringByReplacingOccurrencesOfString:@"[stream_id]" withString:[[NSUUID UUID] UUIDString]];
+    // replacing user_id with random UUID here for testing, please user the real user ID from the system.
+    url = [url stringByReplacingOccurrencesOfString:@"[user_id]" withString:[[NSUUID UUID] UUIDString]];
+    
+    // Fetch the xml from server
+    NSXMLParser *xmlparser = [[NSXMLParser alloc] initWithContentsOfURL:[[NSURL alloc] initWithString:url]];
+    
+    // Or use the hardcoded copy
+    // NSData* vastSoloData = [NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"vast_solo" ofType:@"xml"]];
+    // NSXMLParser *xmlparser = [[NSXMLParser alloc] initWithData:vastSoloData];
+    
+    [xmlparser setDelegate:self];
+    BOOL success = [xmlparser parse];
+    if (success) {
+//        // Print vastDictionary as JSON string to help with debug.
+//        NSError *error;
+//        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:self.vastDictionary
+//                                                           options:NSJSONWritingPrettyPrinted && NSJSONWritingSortedKeys
+//                                                             error:&error];
+//        if (!jsonData) {
+//            NSLog(@"Got an error: %@", error);
+//        } else {
+//            NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+//            NSLog(@"vastDictionary: %@", jsonString);
+//        }
+        _vastReady = YES;
+        NSLog(@"fetchAd: ready");
+    } else {
+        [self alertWithTitle:@"Error" message:@"Failed to fetch vmap." completion:nil];
+    }
+}
+
+- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName attributes:(NSDictionary *)attributeDict
+{
+    NSMutableDictionary* parent = self.currentPointer;
+    self.currentPointer = [attributeDict mutableCopy];
+    
+    if (parent) {
+        [self.currentPointer setObject:parent forKey:@"parent"];
+        NSMutableArray* siblings = [parent objectForKey:elementName];
+        if (!siblings) {
+            [parent setObject:[[NSMutableArray alloc] initWithArray:@[]] forKey:elementName];
+            siblings = [parent objectForKey:elementName];
+        }
+        [siblings addObject:self.currentPointer];
+    }
+    
+    // setting the root
+    if (!self.vastDictionary) {
+        self.vastDictionary = self.currentPointer;
+    }
+}
+
+- (void)parser:(NSXMLParser *)parser foundCDATA:(NSData *)CDATABlock {
```

Comment:
Ah I see, the currentPointer is for XML parsing! Cool. Perhaps we can use a more specific variable name still though

Reply:
I am open to naming. I am bad at names.